{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Unix Basics\n",
    "\n",
    "The command line allows you to input commands, such as creating folders, deleting and copying files and extracting information from files.\n",
    "\n",
    "By the end of this notebook, you will...\n",
    "\n",
    "* Have (maybe installed) and used the command line on your personal laptop\n",
    "* Created files and folders, and looked around a directory, using just the keyboard\n",
    "* Used command line tools like `sed`, `awk`, and `grep` to find replace text, fetch columns, and find words in files\n",
    "\n",
    "* * *\n",
    "\n",
    "## Opening the command line.\n",
    "\n",
    "* **From Mac OS**\n",
    "    * Applications folder, open Utilities and launch Terminal  \n",
    "* **From Linux machine**\n",
    "    * Applications, Accessories and launch Terminal  \n",
    "* **From Windows**\n",
    "    * Determine if you have a 32 or 64 bit version of Windows. \n",
    "        * https://support.microsoft.com/en-us/kb/827218. \n",
    "        * http://cygwin.com/install.html.\n",
    "    * Run setup-x86.exe if you have 32 bit windows. If not Run setup-x86_64.exe\n",
    "    * Install Cygwin and double-click on Cygwin Terminal\n",
    "\n",
    "\n",
    "## Getting Started: Navigating your folders and files\n",
    "You start any terminal session in your \"home area\".  View your \"present working directory\"\n",
    "* `pwd`\n",
    "\n",
    "Your default home area is represented by the character alias `~` (tilde)\n",
    "* `print ~`\n",
    "\n",
    "Change directory  \n",
    "* `cd ~/Desktop`  \n",
    "\n",
    "List all the files in the present working directory using  \n",
    "* `ls`  \n",
    "* `ls .`\n",
    "    \n",
    "Arguments for unix commands  \n",
    "* `man ls`  \n",
    "    \n",
    "Creating a folder  \n",
    "* `mkdir data`  \n",
    "* `mkdir software`  \n",
    "    \n",
    "Change directory into data or software (tab complete or use Up and Down)\n",
    "* `cd da[TAB]`  \n",
    "\n",
    "Change back to the root directory from any subdirectory:\n",
    "* `cd ..`  \n",
    "\n",
    "Create an empty file\n",
    "* `touch emptyfile.txt`  \n",
    "\n",
    "Write some text in it\n",
    "* `echo \"hello world\" > emptyfile.txt`\n",
    "\n",
    "Move or rename a file\n",
    "* `mv emptyfile.txt notempty.txt`  \n",
    "\n",
    "Copy a file\n",
    "* `cp notempty.txt deleteme.txt`  \n",
    "\n",
    "Delete a file\n",
    "* `rm deleteme.txt`  \n",
    "\n",
    "Create a pointer (symlink) to a file\n",
    "* `ln -s notempty.txt pointer`  \n",
    "\n",
    "## File Manipulation: Getting some data from UCSC's Table Browser\n",
    "Go to the UCSC genome browser and the knownGene table (save as knowngene.txt)  \n",
    "* http://hgdownload.cse.ucsc.edu/downloads.html\n",
    "\n",
    "Move knownGene.txt to Desktop\n",
    "* `mv Downloads/knownGene.txt Desktop`  \n",
    "    \n",
    "(*optional*) Secure copy knownGene.txt to TSCC.\n",
    "* `scp Desktop/knownGene.txt  ewyeo@tscc-login2.sdsc.edu:.`  \n",
    "    \n",
    "What's in the file?\n",
    "* `less knownGene.txt`  \n",
    "* `cat knownGene.txt`  \n",
    "\n",
    "How many lines are in the file?\n",
    "* `wc -l knownGene.txt`  \n",
    "    \n",
    "See what's in the first n lines\n",
    "* `head -n 10 knownGene.txt`  \n",
    "    \n",
    "Check if it's indeed n lines (| command)\n",
    "* `less knownGene.txt | wc -l`\n",
    "* `wc -l knownGene.txt`  \n",
    "    \n",
    "What's in the last n lines?\n",
    "* `tail -n 10 knownGene.txt`  \n",
    "    \n",
    "Extract specific columns\n",
    "* `cut -f`  \n",
    "* `paste column1.txt column2.txt > 2columns.txt`\n",
    "\n",
    "How many genes have 3 exons?\n",
    "* `grep -c 'REGEXSEARCHTERM' target.txt`  \n",
    "    \n",
    "How many genes have 1...max # exons?\n",
    "* `sort | uniq -c`  \n",
    "    \n",
    "\n",
    "## Deleting files and file permissions\n",
    "Which user are you logged in as?\n",
    "* `whoami`\n",
    "\n",
    "What groups is that user associated with?\n",
    "* `groups`\n",
    "\n",
    "What is the ownership status of all files in my current directory?\n",
    "* `ls -lrt`\n",
    "* [Interpreting output](http://linuxcommand.org/lts0070.php)  \n",
    "\n",
    "Changing permissions\n",
    "* `chmod 775`\n",
    "\n",
    "The three digits indicate the affected user subset:\n",
    "* Front = Owner  \n",
    "* Middle = Group  \n",
    "* Rear = All Users  \n",
    "\n",
    "The value indicates visibility:  \n",
    "* 7 = May read and write  \n",
    "* 5 = May read but not write  \n",
    "* 0 = May not read or write  \n",
    "\n",
    "Changing Files Recursively\n",
    "* `chmod -R 777 Directory/`\n",
    "* `chmod -R o-rwx ~/`\n",
    "\n",
    "Changing executable nature of files\n",
    "* `chmod +x`\n",
    "\n",
    "Scratch maintenance occurs every 90 days:\n",
    "* `cd important_scratch_dir`\n",
    "* `find . | xargs touch`\n",
    "\n",
    "## Introduction to `awk`\n",
    "\n",
    "`awk` is a command-line tool to \n",
    "\n",
    "Another way to extract all lines\n",
    "\n",
    "```\n",
    "awk -F \"\\t\" '{print;}' knownGene.txt\n",
    "```\n",
    " \n",
    "What if we only wanted one column\n",
    "\n",
    "```\n",
    "awk -F \"\\t\" '{print $8;}' knownGene.txt  | head\n",
    "```\n",
    "\n",
    "What if we wanted the length of genes?\n",
    "\n",
    "```\n",
    "awk -F \"\\t\" '{ len = $5-$4;} {print len;}' knownGene.txt | head\n",
    "```\n",
    "\n",
    "Length of all genes summed?\n",
    "```\n",
    "awk -F \"\\t\" '{ len = $5-$4;} {tot = tot + len;} END {print tot;}' knownGene.txt | head\n",
    "```\n",
    "\n",
    "Don't process the header line (introduction to conditionals)\n",
    "* `awk -F \"\\t\" '{`\n",
    "* `if (FNR==1){`\n",
    "    *`next`\n",
    "* };\n",
    "* tot = tot + $5-$4};\n",
    "* END {print tot;}' knownGene.txt | head`\n",
    " \n",
    "What if you only want the total length of genes in chromosome 1?\n",
    "* `awk -F \"\\t\" '{`\n",
    "    * `if (FNR==1){`\n",
    "        * `next;`\n",
    "    * `};`\n",
    "    * `chr =$2;`\n",
    "    * `if (chr == \"chr1\") {`\n",
    "        * `tot = tot + $5-$4;`\n",
    "    * `}`\n",
    "* `};`\n",
    "* `END {print tot;}' knownGene.txt`\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
