{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* * *\n",
    "\n",
    "#  File Download and Processing Example\n",
    "\n",
    "## In-Class Project : Gather Data From Ensembl biomart\n",
    "\n",
    "#### What is an annotation?\n",
    "An annotation is a flag denoting a special characteristic to a particular base pair or string of base pairs in a DNA sequence.  The most common type of annotation assigns particular chromosomal coordinates to a gene ID (and its associated metadata.)  Genetic sequence is most commonly stored in strings of 'A', 'C', 'G', and 'T' (fasta format).  Annotations are sometimes bundled with sequence (gbk, genbank file) or downloaded separately in memory efficient XML-based formats (gtf).\n",
    "\n",
    "#### Install the BiomaRt package (R package for simple Ensemble queries)\n",
    "[BiomaRt run guide (R)](http://www.ensembl.info/blog/2015/06/01/biomart-or-how-to-access-the-ensembl-data-from-r/)\n",
    "Download R (or Rstudio) and Install\n",
    "Start R (or RStudio).  Using R session terminal...\n",
    "Install biomaRt\n",
    "* `source(\"http://bioconductor.org/biocLite.R\")`\n",
    "* `biocLite(\"biomaRt\")`  \n",
    "      \n",
    "Test install by listing available ensembl data types, or \"marts\":\n",
    "* `library(biomaRt)`\n",
    "* `listEnsembl()`  \n",
    "      \n",
    "List filters\n",
    "* `listFilters(variation)`  \n",
    "      \n",
    "List attributes\n",
    "* `listAttributes(variation)`  \n",
    "      \n",
    "#### Gather Ensembl IDs of Interest\n",
    "Search chromosome hg19 (GRCh37) for the following attributes:\n",
    "* Ensembl gene id\n",
    "* Transcript ID\n",
    "* Aassociated gene name\n",
    "\n",
    "Example R commands:\n",
    "* `library(biomaRt)`\n",
    "* `ensembl = useEnsembl(biomart=\"ensembl\", dataset=\"hsapiens_gene_ensembl\")`\n",
    "* `genes <- getBM(attributes=c('ensembl_gene_id','gene_biotype','hgnc_symbol'), mart = ensembl)\n",
    "* `View genes`\n",
    "\n",
    "#### Convert Ensembl ids to ucsc ids\n",
    "##### Method 1:  (R)\n",
    "[Source](http://genomewiki.ucsc.edu/index.php/Gene_id_conversion)\n",
    "Biomart http://www.biomart.org is probably the best solution if your source ids are from Ensembl:\n",
    "* Click-Path:\n",
    "    * martview (top-right of screen)\n",
    "    * ensembl56 genes\n",
    "* (select your species)\n",
    "    * \"filters\"\n",
    "    * \"gene\"\n",
    "* paste your ids into \"id list limit\"\n",
    "    * \"attributes\"\n",
    "    * \"GENE\"\n",
    "* uncheck \"ensembl transcript id\"\n",
    "* uncheck \"ensembl gene id\" if you want to get rid of it\n",
    "    * \"EXTERNAL\"\n",
    "* check \"HGNC symbol\" (or \"HGNC automatic gene name\" if not human)\n",
    "\"results\"\n",
    "\n",
    "##### Method 2: (mysql)\n",
    "[Source; Pierre Lindenbaum](https://www.biostars.org/p/3570/)\n",
    "* `mysql --user=genome --host=genome-mysql.cse.ucsc.edu -A -D hg18`  \n",
    "\n",
    "For each ID ...\n",
    "* `mysql> select * from ensGene as G,ensGtp as T`\n",
    "* `where G.name=T.transcript  and T.gene=\"ENSG00000215719\" limit 1\\G`     \n",
    "\n",
    "#### Getting 3’ UTR fasta sequences with ensembl gene id as a header\n",
    "* \n",
    "(this will be useful for motif stuff later in day 2 -olga we can see which motifs from cisbp are enriched in 3’UTrs?)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
