{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Day 1: Unix Intro\n",
    "\n",
    "Using the command line. The command line allows you to input commands, such as creating folders, deleting and copying files and extracting information from files.\n",
    "\n",
    "## Opening the command line.\n",
    "### From Mac OS\n",
    "1. Applications folder, open Utilities and launch Terminal\n",
    "### From Linux machine\n",
    "1. Applications, Accessories and launch Terminal\n",
    "### From Windows\n",
    "1. Determine if you have a 32 or 64 bit version of Windows.\n",
    "https://support.microsoft.com/en-us/kb/827218.\n",
    "http://cygwin.com/install.html.\n",
    "2. Run setup-x86.exe if you have 32 bit windows. If not Run setup-x86_64.exe\n",
    "3. Install Cygwin and double-click on Cygwin Terminal\n",
    "   \n",
    "## Navigating your folders and files\n",
    "1. Change directory\n",
    "```\n",
    "cd ~/Desktop\n",
    "```\n",
    "2. List all the files present there using \n",
    "```\n",
    "ls \n",
    "```\n",
    "3. Arguments for unix commands\n",
    "```\n",
    "man ls \n",
    "```\n",
    "4. Creating a folder \n",
    "```\n",
    "mkdir data\n",
    "mkdir software \n",
    "```\n",
    "5. Change directory into data or software (tab complete or use Up and Down)\n",
    "```\n",
    "?\n",
    "```\n",
    "\n",
    "\n",
    "## Getting some data from UCSC's Table Browser\n",
    "1. Go to the UCSC genome browser and the knownGene table (save as knowngene.txt)\n",
    "2. Move knownGene.txt to Desktop\n",
    "```\n",
    "mv Downloads/knownGene.txt Desktop\n",
    "```\n",
    "3. *optional* Secure copy knownGene.txt to TSCC.\n",
    "```\n",
    "scp Desktop/knownGene.txt  ewyeo@tscc-login2.sdsc.edu:.\n",
    "```\n",
    "4. What's in the file?\n",
    "``` \n",
    "less knownGene.txt \n",
    "```\n",
    "5. How many lines are in the file?\n",
    "```\n",
    "wc knownGene.txt\n",
    "```\n",
    "6. See what's in the first n lines\n",
    "```\n",
    "head -n 10 knownGene.txt\n",
    "```\n",
    "7. Check if it's indeed n lines (| command)\n",
    "8. What's in the last n lines?\n",
    "```\n",
    "tail -n 10 knownGene.txt\n",
    "```\n",
    "9. Extract specific columns \n",
    "``` \n",
    "cut -f \n",
    "```\n",
    "10. How many genes have 3 exons?\n",
    "```\n",
    "grep\n",
    "```\n",
    "11. How many genes have 1...max # exons?\n",
    "``` \n",
    "sort | uniq -c\n",
    "```\n",
    "12. Output to a file and open in excel (make bar chart)\n",
    "\n",
    "\n",
    "## Introduction to awk\n",
    "awk is a useful tool \n",
    "\n",
    "1. Another way to extract all lines\n",
    "```\n",
    "awk -F \"\\t\" '{print;}' knownGene.txt\n",
    "```\n",
    "\n",
    "2. What if we only wanted one column\n",
    "```\n",
    "awk -F \"\\t\" '{print $8;}' knownGene.txt  | head\n",
    "```\n",
    "\n",
    "3. What if we wanted the length of genes?\n",
    "```\n",
    "awk -F \"\\t\" '{ len = $5-$4;} {print len;}' knownGene.txt | head\n",
    "```\n",
    "\n",
    "4. Length of all genes summed?\n",
    "```\n",
    "awk -F \"\\t\" '{ len = $5-$4;} {tot = tot + len;} END {print tot;}' knownGene.txt | head\n",
    "```\n",
    "5. Don't process the header line (introduction to conditionals)\n",
    "```\n",
    "awk -F \"\\t\" '{\n",
    "    if (FNR==1){\n",
    "        next\n",
    "    };\n",
    "    tot = tot + $5-$4}; \n",
    "    \n",
    "    END {print tot;}' knownGene.txt | head\n",
    "```\n",
    "6. What if you only want the total length of genes in chromosome 1?\n",
    "```\n",
    "awk -F \"\\t\" '{\n",
    "    if (FNR==1){\n",
    "        next;\n",
    "    };\n",
    "       chr =$2;\n",
    "    if (chr == \"chr1\") {\n",
    "       tot = tot + $5-$4;\n",
    "    }\n",
    "        }; \n",
    "    \n",
    "    END {print tot;}' knownGene.txt \n",
    "```\n",
    "\n",
    "## Deleting files and file permissions\n",
    "\n",
    "## Getting some data from Ensembl biomart\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
