{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Homework 1: File formats and TF binding sites\n",
    "\n",
    "By the end of this homework you will...\n",
    "\n",
    "* Know what common genome annotation formats are (BED, GTF)\n",
    "* Used knowledge of UNIX commands to ask biological questions\n",
    "* Know how to use the UCSC genome browser to download BED files\n",
    "* Loaded `biotools` on TSCC to get access to the `bedtools` program\n",
    "* Used `bedtools` to get promoters of genes, overlap them with transcription factor binding sites \n",
    "* Copied data from TSCC to your laptop\n",
    "* Become more comfortable with reading and navigating documentation and man pages\n",
    "\n",
    "## Log in to TSCC\n",
    "\n",
    "Log in to tscc:\n",
    "\n",
    "```\n",
    "ssh username@tscc.sdsc.edu\n",
    "```\n",
    "\n",
    "## Clone the `biom262-2016` Github Repo\n",
    "\n",
    "1. Read sections 10-12 from Software Carpentry's [Git intro](http://swcarpentry.github.io/git-novice/index.html).\n",
    "\n",
    "You should have a `code` and `notebooks` directory already. If not, create them. Since the `biom262-2016` is essentially a bunch of code, we'll clone this into the `code` directory. Go into the `code` directory and clone the repo:\n",
    "\n",
    "```\n",
    "git clone https://github.com/biom262/biom262-2016.git\n",
    "```\n",
    "\n",
    "Now change directories to `weeks/week01/data`\n",
    "\n",
    "* * *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nbgrader": {
     "grade": false,
     "locked": true,
     "solution": false
    }
   },
   "source": [
    "## What is an annotation?\n",
    "An annotation is a flag denoting a special characteristic to a particular base pair or string of base pairs in a DNA sequence.  The most common type of annotation assigns particular chromosomal coordinates to a gene ID (and its associated metadata.)  Genetic sequence is most commonly stored in strings of 'A', 'C', 'G', and 'T' (fasta format).  Annotations are sometimes bundled with sequence (gbk, genbank file) or downloaded separately in memory efficient formats like GTF and BED.\n",
    "\n",
    "### BED Format: stuff in the genome\n",
    "\n",
    "BED stands for \"Browser Extensible Data\" (informative I know .....) and is a standard format in bioinformatics for describing locations of stuff in the genome the basic format is described [here](https://genome.ucsc.edu/FAQ/FAQformat.html#format1), and here's a minimal example (from [`pybedtools`](https://github.com/daler/pybedtools/blob/master/pybedtools/test/data/a.bed) tests):\n",
    "\n",
    "```\n",
    "chr1\t0\t100\tfeature1\t0\t+\n",
    "chr1\t100\t200\tfeature2\t0\t+\n",
    "chr1\t150\t500\tfeature3\t0\t-\n",
    "chr1\t900\t950\tfeature4\t0\t+\n",
    "```\n",
    "\n",
    "As you can see, the format is:\n",
    "\n",
    "1. Chromosome name\n",
    "2. Start of the feature (0-based, so the first base of the genome is \"0\")\n",
    "3. Stop of the feature (0-based, so not inclusive). This is for computer science reasons, most programming langauges are written such that the \"0th\" item is the first thing. So then the \"100th\" item would then actually be the 101th element. To avoid the \"off by one error\" (a common problem in bioinformatics) then we use 0-based indexing, where `feature1` above starts at base 0 and ends at base 100, and thus is of length 100.\n",
    "4. Name of the feature\n",
    "5. Score of the feature, which is some integer from 0 to 1000. Some bed files come with a dot/period \"`.`\" here instead if it doesn't make sense for them to have a score, but there's programs that will complain that your `bed` file is improperly formatted so I end up using `awk` or something to fill that column with 1000 for every row.\n",
    "6. Strand of the feature. If a strand doesn't make sense (e.g. for DNA methylation) then a dot/period \"`.`\"  is here.\n",
    "\n",
    "### Get Transcription Factor Binding Site BED files from the UCSC Table Browser\n",
    "\n",
    "Go to the [UCSC Genome Browser](http://genome.ucsc.edu/) and click \"Tools > Table Browser\", as shown below.\n",
    "\n",
    "![](table_browser.png \"UCSC Table Browser\")\n",
    "\n",
    "1. Go to the [Table Browser](http://genome.ucsc.edu/cgi-bin/hgTables) \n",
    "2. Choose \"clade: Mammal,\" \"genome: Human,\" and \"assembly: Feb. 2009 (GRCh37/hg19)\"\n",
    "3. Choose \"group: Regulation\", \"track: Txn Fac ChIP V2\"\n",
    "3. Use the default table (there are no others anyway)\n",
    "4. Put \"chr22\" for the Region\n",
    "5. Choose \"output format: BED - browser extensible data\". \n",
    "4. Save as \"`tf.bed`\"\n",
    "4. Click \"get output\".\n",
    "4. Do one bed record per gene (They really mean per \"item\", not a whole gene)\n",
    "5. Click \"get BED\"\n",
    "\n",
    "What does this file look like? Is it similar to what we had before? Let's use `head` to look at the first few lines, which by default is 10 lines.\n",
    "\n",
    "\n",
    "Hint: Remember that the dollar sign (\"`$`\") indicates the shell command prompt that you're in, and shouldn't be included when you're copying commands.\n",
    "\n",
    "```\n",
    "$ head ~/Downloads/tf.bed     \n",
    "chr22\t16166497\t16166741\tCTCF\t186\t.\t16166497\t16166741\t0\t1\t244,\t0,\n",
    "chr22\t16201947\t16202317\tCTCF\t603\t.\t16201947\t16202317\t0\t1\t370,\t0,\n",
    "chr22\t16201988\t16202252\tYY1_(C-20)\t63\t.\t16201988\t16202252\t0\t1\t264,\t0,\n",
    "chr22\t16202021\t16202231\tRad21\t92\t.\t16202021\t16202231\t0\t1\t210,\t0,\n",
    "chr22\t16202128\t16202242\tE2F6_(H-50)\t269\t.\t16202128\t16202242\t0\t1\t114,\t0,\n",
    "chr22\t16205233\t16205683\tSETDB1\t356\t.\t16205233\t16205683\t0\t1\t450,\t0,\n",
    "chr22\t16325696\t16325900\tNRSF\t36\t.\t16325696\t16325900\t0\t1\t204,\t0,\n",
    "chr22\t16560670\t16560910\tMafK_(ab50322)\t102\t.\t16560670\t16560910\t0\t1\t240,\t0,\n",
    "chr22\t16872258\t16872628\tCTCF\t279\t.\t16872258\t16872628\t0\t1\t370,\t0,\n",
    "chr22\t16872319\t16872563\tSMC3_(ab9263)\t309\t.\t16872319\t16872563\t0\t1\t244,\t0\n",
    "```\n",
    "\n",
    "### Copy `tf.bed` to TSCC\n",
    "\n",
    "#### Mac/Linux\n",
    "\n",
    "To copy data between computers, you can use \"secure copy\" or \"`scp`\". Just like with `cp` you do `cp filename newplace` you can do the same thing, but you have to specify the server name, and where on the server you want to put it. \n",
    "\n",
    "*Note: you are running this command from your **laptop**, because you need to be able to specify a unique address/URL for one of the file locations. Unless you have command line login and a static IP for your laptop (which takes a lot of work and is therefore unlikely), then you cannot run `scp` from TSCC to copy files to/from your laptop. You must always run `scp` from your laptop if you want to move files to/from a server.*\n",
    "\n",
    "```\n",
    "$ scp tf.bed username@tscc.sdsc.edu:~/code/biom262-2016/weeks/week01/data\n",
    "```\n",
    "\n",
    "Written another more generically, the `scp` command looks like\n",
    "\n",
    "```\n",
    "$ scp localfile username@server:/path/to/new/place/on/server\n",
    "```\n",
    "\n",
    "#### Windows\n",
    "\n",
    "(Instructions compiled from [Cornell IT](http://www.it.cornell.edu/services/managed_servers/howto/file_transfer/fileputty.cfm) and [Analyzing Next-Gen Seq (ANGUS) data workshop](http://ged.msu.edu/angus/tutorials/using-putty-on-windows.html))\n",
    "\n",
    "Install [PuTTY](http://www.chiark.greenend.org.uk/~sgtatham/putty/download.html). It doesn't need to be installed so you can put the file on your desktop or wherever you like to put your programs. You'll need both the `putty.exe` and `pscp.exe` files.\n",
    "\n",
    "Open the windows command line \n",
    "\n",
    "![](http://www.it.cornell.edu/catc/cms/services/managed_servers/howto/file_transfer/images/filewindowsstart_1.jpg)\n",
    "![](http://www.it.cornell.edu/catc/cms/services/managed_servers/howto/file_transfer/images/filewindowscmd_1.jpg)\n",
    "\n",
    "At the command prompt, you can now enter same command as for Mac/Linux, but using `pscp` instead of `scp`.\n",
    "\n",
    "```\n",
    "c:\\>pscp tf.bed username@tscc.sdsc.edu:~/code/biom262-2016/weeks/week01/data\n",
    "```\n",
    "\n",
    "Written another more generically, the `pscp` command looks like\n",
    "\n",
    "```\n",
    "c:\\>pscp localfile username@server:/path/to/new/place/on/server\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 1: Get NFKB binding sites only\n",
    "\n",
    "Use your UNIX skillz to filter the `tf.bed` file for only the NFKB TFs. Save this file as `tf.nkfb.bed`. This file should have 702 lines."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gencode gene annotations\n",
    "\n",
    "We are interested in which gene's promoters these transcription factors bind to. We will use the [GENCODE](http://www.gencodegenes.org/) gene annotation, which is an aggregate of several gene annotation groups (ENSEMBL, HAVANA) to create a comprehensive set of human, and recently, mouse, gene annotations. We downloaded the annotations using [GENCODE v19](http://www.gencodegenes.org/releases/19.html), the last release using hg19, which as of writing in January 2016 has all the ENCODE transcription factor binding sites.\n",
    "\n",
    "If you're interested in the difference between the GENCODE gene annotations and RefSeq genes, check out [this](http://www.biomedcentral.com/1471-2164/16/S8/S2) paper.\n",
    "\n",
    "For this assignment, we have provided the gene annotations and sequence for human chromosome 22 (one of the smallest chromosomes), which are located in `biom262-2016/weeks/week01/data`\n",
    "\n",
    "#### In case you are curious ...\n",
    "This information is just so you know how grabbing just chromosome 22 for the different file formats was accomplished, here it is. It is not graded material. \n",
    "\n",
    "To get the `chr22` data from the gtf file, we can `grep` for lines that start with `chr22`, where the up-carat/power symbol \"`^`\" indicates the start of the line.\n",
    "\n",
    "```\n",
    "$ gunzip gencode.v24.basic.annotation.gtf.gz\n",
    "$ grep '^chr22' gencode.v19.annotation.gtf > gencode.v19.annotation.chr22.gtf\n",
    "```\n",
    "\n",
    "To get the fasta file, we first had to index it using `samtools faidx`, and then we could use the same program to grab chromosome 22.\n",
    "\n",
    "```\n",
    "$ gunzip GRCh37.p13.genome.fa.gz \n",
    "$ samtools faidx GRCh37.p13.genome.fa.gz \n",
    "$ samtools faidx GRCh37.p13.genome.fa.gz chr22 > GRCh37.p13.chr22.fa \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fasta and GTF formats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Fasta\n",
    "\n",
    "The FASTA format is one of the oldest in bioinformatics and the simplest. The concept is this:\n",
    "\n",
    "```\n",
    ">sequence1_name\n",
    "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGCCCCCCCCCC\n",
    "CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGGGGGGGGGG\n",
    "GGGGGGGGGG\n",
    ">sequence2_name\n",
    "CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCGGGGGGGGGG\n",
    "GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGCCCCCCCCCC\n",
    "CCCCCCCCCC\n",
    "```\n",
    "\n",
    "Where a sequence entry starts with a greater than sign \"`>`\" and any number of sequence lines can follow. The end of the sequence is indicated by the start of the next one (or the end of the file!). The example above shows two 110-base long sequences, with the first two rows of the sequence are exactly the same length (50bp), with a final hanging sequence (10bp). Some programs require that the lines be the same length, so if you're creating FASTA files by hand, it's recommended to use [BioPython's SeqIO](http://biopython.org/wiki/SeqIO) to generate them so they're correctly formatted.\n",
    "\n",
    "\n",
    "We can use `head` to look at the first few lines (10 by default) of each of the files. The fasta file is kind of boring for the first few lines:\n",
    "\n",
    "\n",
    "```\n",
    "$ head GRCh37.p13.chr22.fa\n",
    ">chr22\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "```\n",
    "\n",
    "One question is, how many bases are in this file? Well, we'll need to ignore all rows of the file that start with \"`>`\", which we can do using `grep -v`, where \"`-v`\" *in**v**erts* the match and returns the non-matching lines. Here's the excerpt from `man grep`:\n",
    "\n",
    "     -v, --invert-match\n",
    "             Selected lines are those not matching any of the specified pat-\n",
    "             terns.\n",
    "\n",
    "To do this, we'll use `grep` to search for all lines that start with (startwith=\"`^`\") the character \"`>`\". Remember that the greater than sign is a special character in unix that indicates to save the output to a file with that name, so we need to put quotes around the \"regular expression\" (aka search term) we are searching with to protect it from getting evaluated by the UNIX command line.\n",
    "\n",
    "```\n",
    "$ grep -v '^>' GRCh37.p13.chr22.fa | head\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
    "```\n",
    "\n",
    "Let's make sure we've only removed the first line by looking at the line count.\n",
    "\n",
    "```\n",
    "wc -l \n",
    " 855078 GRCh37.p13.chr22.fa\n",
    "\n",
    "```\n",
    "\n",
    "And this command:\n",
    "\n",
    "```\n",
    "$ grep -v '^>' GRCh37.p13.chr22.fa | wc -l\n",
    "  855077\n",
    "```\n",
    "\n",
    "So yes, we've removed exactly one line. Now we're ready to count the number of bases. Remember that the \"`-l`\" flag on \"`wc`\" indicated to count only the number of lines\n",
    "\n",
    "```\n",
    "$ grep -v '^>' GRCh37.p13.chr22.fa | wc   \n",
    "  855077  855077 52159643\n",
    "```\n",
    "\n",
    "So that means there are 52,159,643 bases in chromosome 22. Cool!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nbgrader": {
     "grade": false,
     "locked": true,
     "solution": false
    }
   },
   "source": [
    "##### GTF\n",
    "\n",
    "Like the BED format, GTF encodes \"stuff in the genome,\" but allows for hierarchy in the relationships between items and for additional annotations of the items. For example, GTFs can encode genes, which have transcripts as children, which have exons as children, which have coding sequences (CDS), untranslated regions (UTRs), stop codons, and more as children. This is not possible in BED files.\n",
    "\n",
    "Read the GENCODE description of their [GTF](http://www.gencodegenes.org/data_format.html) format, which describes their columns and convenient anntations like whether the gene is protein-coding, the reading frame of coding sequences (CDS), and at which \"level\" a gene is annotated - verified, manually, or computationally.\n",
    "\n",
    "Here is a screenshot of the GTF file from the GENCODE website.\n",
    "\n",
    "![](gtf_example.png \"GTF Example\")\n",
    "\n",
    "As you can see, it is *similar* but *different* than the bed file. Besides the columns, a key difference is that GTF files are 1-based but BED files are 0-based. Thus, if we have a feature which in BED-land:\n",
    "\n",
    "```\n",
    "chr1\t0\t100\tfeature1\t0\t+\n",
    "```\n",
    "\n",
    "Then in the GTF, we have:\n",
    "\n",
    "```\n",
    "chr1\tsource\tgene\t1\t100\tfeature1\t0\t+\t.\n",
    "```\n",
    "\n",
    "The GTF file is slightly easier to understand because the feature starts at 1 (and is 1-based so \"1\" is truly the first nucleotide) truly ends on base 99, and there is no more feature on base 100. How is this different for the negative strand?\n",
    "\n",
    "For a bed file with the following 0-based coordinates, what would be the coordinates in the 1-based gtf file?\n",
    "\n",
    "```\n",
    "chr1\t150\t500\tfeature3\t0\t-\n",
    "```\n",
    "\n",
    "It'd be the same - the start is only changed. Why is that?\n",
    "\n",
    "```\n",
    "chr1\tsource\tgene\t151\t500\tfeature3\t0\t-\n",
    "```\n",
    "\n",
    "\n",
    "Now let's look at GTF files on the command line. What if we do `head` on just the first two lines?\n",
    "\n",
    "```\n",
    "head -n 2 gencode.v19.annotation.chr22.gtf\n",
    "chr22\tHAVANA\tgene\t16062157\t16063236\t.\t+\t.\tgene_id \"ENSG00000233866.1\"; transcript_id \"ENSG00000233866.1\"; gene_type \"lincRNA\"; gene_status \"KNOWN\"; gene_name \"LA16c-4G1.3\"; transcript_type \"lincRNA\"; transcript_status \"KNOWN\"; transcript_name \"LA16c-4G1.3\"; level 2; havana_gene \"OTTHUMG00000140195.1\";\n",
    "chr22\tHAVANA\ttranscript\t16062157\t16063236\t.\t+\t.\tgene_id \"ENSG00000233866.1\"; transcript_id \"ENST00000424770.1\"; gene_type \"lincRNA\"; gene_status \"KNOWN\"; gene_name \"LA16c-4G1.3\"; transcript_type \"lincRNA\"; transcript_status \"KNOWN\"; transcript_name \"LA16c-4G1.3-001\"; level 2; tag \"basic\"; havana_gene \"OTTHUMG00000140195.1\"; havana_transcript \"OTTHUMT00000276574.1\";\n",
    "```\n",
    "\n",
    "\n",
    "Oh that's annoying - GTF files are very wwwwwwwwwwwwwwiiiiiiiiiiiiiidddddddddddddddddddddeeeeeeeeeeeeeeeeeeeeeeeee and so the line wraps around and makes it are difficult to view in the terminal and line up columns. So we don't want to use `head`. Instead we'll use `less` with the `-S` flag. Check out `man less` to see **all** the options (there's a ton). Here's an excerpt from `man less` to show the function of `-S`:\n",
    "\n",
    "       -S or --chop-long-lines\n",
    "              Causes  lines  longer than the screen width to be chopped rather\n",
    "              than folded.  That is, the portion of a long line that does  not\n",
    "              fit  in  the  screen width is not shown.  The default is to fold\n",
    "              long lines; that is, display the remainder on the next line.\n",
    "\n",
    "\n",
    "```\n",
    "$ less -S gencode.v19.annotation.chr22.gtf\n",
    "```\n",
    "\n",
    "It's nice because you can use the left and right arrow keys to navigate to different columns.\n",
    "\n",
    "Here's a screenshot of the output:\n",
    "\n",
    "![](gencode_v19_chr22_less-S.png \"Example with 'less -S' to chop instead of wrap lines\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 2: Extract only the genes from the GTF file\n",
    "\n",
    "Use what you've learned about UNIX so far to get only the rows of `gencode.v19.annotation.chr22.gtf` that contain features of type \"`transcript`\", saving to a file called `gencode.v19.annotation.chr22.transcript.gtf`. Your file should contain `4459` lines."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Use Bedtools to get promoters and overlap with TF binding sites\n",
    "\n",
    "[Bedtools](http://bedtools.readthedocs.org/en/latest/) is a \"swiss-army knife of tools for a wide-range of genomics analysis tasks\". In particular, `bedtools` ***excels*** at \"genome algebra\", the adding and subtracting of genomic regions together. It is used ***EXTENSIVELY*** in bioinformatics and there are few projects that can live without it (or creating their own personal version but who wants to do that?)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load `biotools` on TSCC which includes Bedtools\n",
    "\n",
    "First, try running `bedtools`. You should see this:\n",
    "\n",
    "```\n",
    "$ bedtools\n",
    "bash: bedtools: command not found\n",
    "```\n",
    "\n",
    "This is because your shell doesn't know anything about the command `bedtools`. To load `bedtools` and other bioinformatics tools to your TSCC account, do\n",
    "\n",
    "```\n",
    "module load biotools\n",
    "```\n",
    "\n",
    "This command has no output. To make sure we have `bedtools` available, use `which` to see the full path to `bedtools`:\n",
    "\n",
    "```\n",
    "$ which bedtools\n",
    "/opt/biotools/bedtools/bin/bedtools\n",
    "```\n",
    "\n",
    "(This is a TSCC-specific thing that the nice system administrators have set up for us but is not a general thing you can do on all servers. Maybe other nice sysadmins on other clusters do this but it is not guaranteed)\n",
    "\n",
    "To see all available modules to load, do `module avail`:\n",
    "\n",
    "```\n",
    "[obotvinnik@tscc-login2 ~]$ module avail\n",
    "\n",
    "-------------------------- /opt/modulefiles/applications/.gnu ---------------------------\n",
    "atlas/3.10.2(default)     hdf5/1.8.14(default)      scalapack/2.0.2(default)\n",
    "boost/1.55.0(default)     lapack/3.5.0(default)     slepc/3.5.3(default)\n",
    "fftw/2.1.5                netcdf/3.6.2              sprng/2.0b(default)\n",
    "fftw/3.3.4(default)       netcdf/4.3.2(default)     sundials/2.5.0(default)\n",
    "gsl/1.16(default)         parmetis/4.0.3(default)   superlu/3.3(default)\n",
    "hdf4/2.10(default)        petsc/3.5.2(default)      trilinos/11.12.1(default)\n",
    "\n",
    "------------------------------- /opt/modulefiles/mpi/.gnu -------------------------------\n",
    "mvapich2_ib/2.1rc2(default) openmpi_ib/1.8.4(default)\n",
    "\n",
    "------------------------- /opt/modulefiles/applications/.intel --------------------------\n",
    "atlas/3.10.2(default)     lapack/3.5.0(default)     scalapack/2.0.2(default)\n",
    "boost/1.55.0(default)     mxml/2.9(default)         slepc/3.5.3(default)\n",
    "fftw/2.1.5                netcdf/3.6.2              sprng/2.0b(default)\n",
    "fftw/3.3.4(default)       netcdf/4.3.2(default)     sundials/2.5.0(default)\n",
    "gsl/1.16(default)         papi/5.4.1(default)       superlu/3.3(default)\n",
    "hdf4/2.10(default)        parmetis/4.0.3(default)   tau/2.23(default)\n",
    "hdf5/1.8.14(default)      pdt/3.20(default)         trilinos/11.12.1(default)\n",
    "ipm/2.0.3(default)        petsc/3.5.2(default)\n",
    "\n",
    "------------------------------ /opt/modulefiles/mpi/.intel ------------------------------\n",
    "mvapich2_ib/2.1rc2(default) openmpi_ib/1.8.4(default)\n",
    "\n",
    "---------------------------- /usr/share/Modules/modulefiles -----------------------------\n",
    "dot              module-info      null             rocks-openmpi_ib\n",
    "module-git       modules          rocks-openmpi    use.own\n",
    "\n",
    "----------------------------------- /etc/modulefiles ------------------------------------\n",
    "openmpi-x86_64\n",
    "\n",
    "------------------------------ /opt/modulefiles/compilers -------------------------------\n",
    "cilk/5.4.6(default)           intel/2015.2.164\n",
    "cmake/3.2.1(default)          mono/3.12.0(default)\n",
    "gnu/4.9.2(default)            pgi/14.9(default)\n",
    "guile/2.0.11(default)         python/1(default)\n",
    "intel/2013_sp1.2.144(default) upc/2.20.0(default)\n",
    "\n",
    "----------------------------- /opt/modulefiles/applications -----------------------------\n",
    "abyss/1.5.2(default)        fsa/1.15.9(default)         mpi4py/1.3.1(default)\n",
    "amber/14(default)           gamess/2014.12(default)     namd/2.10(default)\n",
    "apbs/1.3(default)           gaussian/09.D.01(default)   namd/2.9\n",
    "bbcp/14.09.02.00.0(default) globus/5.2.5                nwchem/6.5(default)\n",
    "bbftp/3.2.1(default)        gmp/6.0.0a(default)         octave/3.8.2(default)\n",
    "beagle/2.1(default)         gnutools/2.69(default)      polymake/2.13.1(default)\n",
    "beast/1.8.0                 gromacs/5.0.4(default)      R/3.2.1(default)\n",
    "beast/1.8.1(default)        idl/8.4(default)            rapidminer/6.1.0(default)\n",
    "beast2/2.1.3(default)       jags/3.4.0(default)         scipy/2.7(default)\n",
    "bioroll/6.2(default)        lammps/20141209(default)    siesta/3.2.5(default)\n",
    "biotools/1(default)         matlab/2013a                stata/13.1(default)\n",
    "blcr/0.8.5(default)         matlab/2013b                vasp/4.6\n",
    "cp2k/2.5.1(default)         matlab/2014a                vasp/5.2.12\n",
    "cpmd/3.17.1(default)        matlab/2014b(default)       vasp/5.2.12.gamma\n",
    "cuda/6.5.19(default)        mkl/11.1.2.144(default)     vasp/5.3.5(default)\n",
    "ddt/4.2.2(default)          mpc/1.0.3(default)          vtk/6.1.0(default)\n",
    "eigen/3.2.3(default)        mpfr/3.1.2(default)         weka/3.7.12(default)\n",
    "```\n",
    "\n",
    "To see everything that's loaded, look in `/opt/biotools/`:\n",
    "\n",
    "```\n",
    "$ ls /opt/biotools/\n",
    "bamtools   blat       cufflinks  GenomeAnalysisTK  miRDeep2  randfold    spades       trinity\n",
    "bedtools   bowtie     dendropy   gmap_gsnap        miso      rseqc       squid        velvet\n",
    "biopython  bowtie2    edena      htseq             picard    samtools    stacks       ViennaRNA\n",
    "bismark    bwa        fastqc     idba-ud           plink     soapdenovo  tophat\n",
    "blast      bx-python  fastx      matt              pysam     SOAPsnp     trimmomatic\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 2: Use `bedtools flank` to get the promoters\n",
    "\n",
    "Get 2000bp upstream from each gene. The genome we are using is \"`hg19`\". Save this file as `gencode.v19.annotation.chr22.transcript.promoter.gtf`\n",
    "\n",
    "Go to the documentation for [bedtools](http://bedtools.readthedocs.org/en/latest/) and find the entry for \"`bedtools flank`\".\n",
    "`bedtools flank -h` is also informative (below)\n",
    "\n",
    "```\n",
    "$ bedtools flank -h\n",
    "\n",
    "Tool:    bedtools flank (aka flankBed)\n",
    "Version: v2.22.1\n",
    "Summary: Creates flanking interval(s) for each BED/GFF/VCF feature.\n",
    "\n",
    "Usage:   bedtools flank [OPTIONS] -i <bed/gff/vcf> -g <genome> [-b <int> or (-l and -r)]\n",
    "\n",
    "Options: \n",
    "        -b      Create flanking interval(s) using -b base pairs in each direction.\n",
    "                - (Integer) or (Float, e.g. 0.1) if used with -pct.\n",
    "\n",
    "        -l      The number of base pairs that a flank should start from\n",
    "                orig. start coordinate.\n",
    "                - (Integer) or (Float, e.g. 0.1) if used with -pct.\n",
    "\n",
    "        -r      The number of base pairs that a flank should end from\n",
    "                orig. end coordinate.\n",
    "                - (Integer) or (Float, e.g. 0.1) if used with -pct.\n",
    "\n",
    "        -s      Define -l and -r based on strand.\n",
    "                E.g. if used, -l 500 for a negative-stranded feature, \n",
    "                it will start the flank 500 bp downstream.  Default = false.\n",
    "\n",
    "        -pct    Define -l and -r as a fraction of the feature's length.\n",
    "                E.g. if used on a 1000bp feature, -l 0.50, \n",
    "                will add 500 bp \"upstream\".  Default = false.\n",
    "\n",
    "        -header Print the header from the input file prior to results.\n",
    "\n",
    "Notes: \n",
    "        (1)  Starts will be set to 0 if options would force it below 0.\n",
    "        (2)  Ends will be set to the chromosome length if requested flank would\n",
    "        force it above the max chrom length.\n",
    "        (3)  In contrast to slop, which _extends_ intervals, bedtools flank\n",
    "        creates new intervals from the regions just up- and down-stream\n",
    "        of your existing intervals.\n",
    "        (4)  The genome file should tab delimited and structured as follows:\n",
    "\n",
    "        <chromName><TAB><chromSize>\n",
    "\n",
    "        For example, Human (hg19):\n",
    "        chr1    249250621\n",
    "        chr2    243199373\n",
    "        ...\n",
    "        chr18_gl000207_random   4262\n",
    "\n",
    "Tips: \n",
    "        One can use the UCSC Genome Browser's MySQL database to extract\n",
    "        chromosome sizes. For example, H. sapiens:\n",
    "\n",
    "        mysql --user=genome --host=genome-mysql.cse.ucsc.edu -A -e \\\n",
    "        \"select chrom, size from hg19.chromInfo\"  > hg19.genome\n",
    "````\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 3: Use `bedtools intersect` to overlap TFs with \n",
    "\n",
    "Take a look at the diagrams in the documentation for `bedtools intersect` and think about how it can answer these questions:\n",
    "\n",
    "* Given locations of genome methylation, which genes does it overlap?\n",
    "* Given locations of RBP binding, which exons does it overlap?\n",
    "* Given two CHIP-Seq experiments, which peaks are consistent between them?\n",
    "\n",
    "Use `bedtools intersect` to find which promoters overlap with the NFKB binding sites. Call this file `gencode.v19.annotation.chr22.transcript.promoter.nfkb.gtf`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "#### Use `bedtools getfasta` to extract sequences\n",
    "\n",
    "Read the documentation for [`bedtools getfasta`](http://bedtools.readthedocs.org/en/latest/content/tools/getfasta.html) and figure out how to  request the sequences in `fasta` format for the peaks which overlap in gene promoters\n",
    "\n",
    "Save the file as `knownGene_exons.fasta`.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Not required: HOMER\n",
    "\n",
    "IF you want to use a program to get the binding motif of a transcription factor, you can do that with a `bed` file, the genome `fasta` file, and a motif finding program like [HOMER](http://homer.salk.edu/)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
